core/src/main/java/org/infinispan/container/versioning/irac/DefaultIracTombstoneManager.java

```diff
@@ -232,117 +267,167 @@ public boolean contains(IracTombstoneInfo tombstone) {
-         Map<Address, BackPressure<IracTombstonePrimaryCheckCommand, Void>> staleTombstones = new HashMap<>(tombstoneMap.size());
-         for (IracTombstoneInfo tombstone : tombstoneMap.values()) {
-            DistributionInfo info = distributionManager.getCacheTopology().getSegmentDistribution(tombstone.getSegment());
-            if (!info.isWriteOwner()) {
-               // topology changed, no longer an owner
-               removeTombstone(tombstone);
-               continue;
-            }
-            // if we are a backup owner and the key is present in IracManager, skip checking for the stale tombstone;
-            // it means the primary owner hasn't sent the update to the remote site.
-            if (iracManager.running().containsKey(tombstone.getKey())) {
-               continue;
-            }
-            if (!info.isPrimary()) {
-               BackPressure<IracTombstonePrimaryCheckCommand, Void> backPressure = staleTombstones.get(info.primary());
-               if (backPressure == null) {
-                  backPressure = new BackPressure<>(commandsFactory.buildIracTombstonePrimaryCheckCommand(batchSize), null);
-                  staleTombstones.put(info.primary(), backPressure);
-               }
-               if (backPressure.element.addTombstone(tombstone) == batchSize) {
-                  CompletionStage<Void> rsp;
-                  if (backPressure.delay == null) {
-                     rsp = rpcManager.invokeCommand(info.primary(), backPressure.element, ignoreLeavers(), rpcOptions);
-                  } else {
-                     ReplicableCommand rCmd = backPressure.element;
-                     rsp = backPressure.delay.thenComposeAsync(unused -> rpcManager.invokeCommand(info.primary(), rCmd, ignoreLeavers(), rpcOptions), blockingExecutor);
-                  }
-                  staleTombstones.put(info.primary(), new BackPressure<>(commandsFactory.buildIracTombstonePrimaryCheckCommand(batchSize), rsp));
-               }
-               continue;
-            }
-            stage.dependsOn(new CleanupTask(tombstone).checkRemoteSites());
-         }
+      CompletionStageObservable cso = new CompletionStageObservable();
 
-         //sending any pending stale tombstones
-         for (Map.Entry<Address, BackPressure<IracTombstonePrimaryCheckCommand, Void>> entry : staleTombstones.entrySet()) {
-            BackPressure<IracTombstonePrimaryCheckCommand, Void> backPressure = entry.getValue();
-            if (backPressure.element.isEmpty()) {
-               continue;
-            }
-            if (backPressure.delay == null) {
-               rpcManager.sendTo(entry.getKey(), backPressure.element, DeliverOrder.NONE);
-            } else {
-               backPressure.delay.thenRunAsync(() -> rpcManager.sendTo(entry.getKey(), backPressure.element, DeliverOrder.NONE), blockingExecutor);
-            }
+      Flowable.fromIterable(tombstoneMap.values())
+            .groupBy(this::classifyTombstone, t -> t)
```

Comment:
Okay, sorry but this whole part needs to be rewritten. Unfortunately, groupBy is a deceptively hard operator to get right. The MOST important thing is the next method must subscribe to all possible groups. So when using `concatMap` you need to use `concatMapEager` or `flatMap` providing the parallel subscription value equal to the maximum number of possible groups. If not, the operator can stall causing very hard to detect errors.

I will update with a separate message of what this code should be shortly after review submission.